addons/account/models/account_move.py

```diff
@@ -2507,13 +2507,9 @@ def _reverse_moves(self, default_values_list=None, cancel=False):
                     lambda x: (x.account_id.reconcile or x.account_id.internal_type == 'liquidity')
                               and not x.reconciled
                 )
-                for line in lines:
-                    counterpart_lines = reverse_move.line_ids.filtered(
-                        lambda x: x.account_id == line.account_id
-                                  and x.currency_id == line.currency_id
-                                  and not x.reconciled
-                    )
-                    (line + counterpart_lines).with_context(move_reverse_cancel=cancel).reconcile()
+                for line, counterpart_line in zip(lines, reverse_move.line_ids.filtered(lambda x: x.account_id.reconcile
+                                                                                        or x.account_id.internal_type == 'liquidity')):
+                    (line + counterpart_line).with_context(move_reverse_cancel=cancel).reconcile()
```

Comment:
You don't need to filter anything (duplicated code)
```python
assert len(move.line_ids) == len(reverse_move.line_ids)
for line, reverse_line in zip(move.line_ids, reverse_move.line_ids):
    assert line.account_id == reverse_line.account_id
    if line.account_id.reconcile or line.account_id.internal_type == 'liquidity':
        (line + reverse_line).with_context(move_reverse_cancel=cancel).reconcile()
```

Also, the added assertions are not guaranteed. For instance, the account can change on tax repartitions lines (which seems...weird for a cancel?)
https://github.com/odoo/odoo/blob/af01dac8e35fe814b1d42a1620ff766908425b33/addons/account/models/account_move.py#L2434-L2438

I imagine other lines can be added or removed with cash basis or stock features
____________________
What I propose, is doing something similar as what was done before, but based on the account, not on the lines.

```python
group = defaultdict(list)
for line in move.line_ids + reverse_move.line_ids:
    group[line.account_id].append(line.id)
for account, line_ids in group.items():
    if account.reconcile or account.internal_type == 'liquidity':
        self.env['account.move.line'].browse(line_ids).reconcile()

Reply:
You are right indeed, with the anglosaxon line, the reverse of a sale invoice can give a move with less line with the anglosaxon accounting.
Nice catch, we didn't thought of it.

Same for the different accounts. (but it's not weird to have different accounts. When you reverse an invoice, you get a credit note and thus you can get different account)